enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj

```diff
@@ -14,6 +14,23 @@
     [toucan.db :as db]
     [toucan.hydrate :refer [hydrate]]))
 
+(defn collection-set-for-user
+  "Given an optional user ID, find the transitive set of all Collection IDs which are either:
+  (a) global (ie. no one's personal collection);
+  (b) owned by this user (when user ID is non-nil); or
+  (c) descended from one of the above."
+  [user-or-nil]
+  (let [roots    (db/select ['Collection :id :location :personal_owner_id] :location "/")
+        unowned  (remove :personal_owner_id roots)
+        owned    (when user-or-nil
+                   (filter #(= user-or-nil (:personal_owner_id %)) roots))
```

Comment:
You can do this directly in the query. Something like this for example

```clj
(db/select [Collection ...] :location "/" {:where [:or [:= :personal_owner_id user-id] [:= personal_owner_id nil]]})
```

(You'll probably want to adjust this a bit to generate something more concise when the user ID is `nil` tho)

It's definitely better to filter stuff database-side whenever we can

Reply:
I guess actually you would have to do `db/select` since you need the `:id` and the `:location` to pass to `descendant-ids`. But you could still get away with a single `db/select` and not have to do the `concat`, `filter`, or `remove` stuff